libraries/cbor/src/macros.rs

```diff
@@ -12,6 +12,127 @@
+/// in an optimized manner, consuming the input map.
+///
+/// It takes as input a `BTreeMap` as well as a list of identifiers and keys, and generates code
+/// that assigns the corresponding values to new variables using the given identifiers. Each of
+/// these variables has type `Option<Value>`, to account for the case where keys aren't found.
+///
+/// **Important:** Keys passed to the `read_cbor_map!` macro **must be sorted** in increasing order.
+/// If not, the algorithm can yield incorrect results, such a assigning `None` to a variable even if
+/// the corresponding key existed in the map. **No runtime checks** are made for this in the
+/// `read_cbor_map!` macro, in order to avoid overhead at runtime. However, assertions that keys are
+/// sorted are added in `cfg(test)` mode, so that unit tests can verify ahead of time that the keys
+/// are indeed sorted. This macro is therefore **not suitable for dynamic keys** that can change at
+/// runtime.
+///
+/// Semantically, provided that the keys are sorted as specified above, the following two snippets
+/// of code are equivalent, but the `read_cbor_map!` version is more optimized, as it doesn't
+/// re-balance the `BTreeMap` for each key, contrary to the `BTreeMap::remove` operations.
+///
+/// ```rust
+/// # extern crate alloc;
+/// # #[macro_use]
+/// # extern crate cbor;
+/// #
+/// # fn main() {
+/// #     let map = alloc::collections::BTreeMap::new();
+/// read_cbor_map! {
+///     map,
+///     x @ 1,
+///     y @ 2,
+/// };
+/// # }
+/// ```
+///
+/// ```rust
+/// # extern crate alloc;
+/// # #[macro_use]
+/// # extern crate cbor;
+/// #
+/// # fn main() {
+/// #     let mut map = alloc::collections::BTreeMap::<cbor::KeyType, _>::new();
+/// let x: Option<cbor::Value> = map.remove(&cbor_unsigned!(1));
+/// let y: Option<cbor::Value> = map.remove(&cbor_unsigned!(2));
+/// # }
+/// ```
+#[macro_export]
+macro_rules! read_cbor_map {
+    ( $map:expr, $( $variable:ident @ $key:expr, )+ ) => {
+        // A pre-requisite for this algorithm to work is that the keys to extract from the map are
+        // sorted.
+        #[cfg(test)]
+        test_ordered_keys!($( $key, )+);
+
+        use $crate::values::{IntoCborKey, Value};
+        use $crate::macros::read_cbor_map_peek_value;
+
+        let mut it = $map.into_iter().peekable();
+        $(
+        let $variable: Option<Value> = read_cbor_map_peek_value(&mut it, $key.into_cbor_key());
+        )+
+    };
+}
+
+/// This function is an internal detail of the `read_cbor_map!` macro, but has public visibility so
+/// that users of the macro can use it.
+///
+/// The logic is separated into its own function to reduce binary size, as otherwise the logic
+/// would be inlined for every use case. As of June 2020, this saves ~40KB of binary size for the
+/// CTAP2 application of OpenSK.
+pub fn read_cbor_map_peek_value(
+    it: &mut Peekable<btree_map::IntoIter<KeyType, Value>>,
```

Comment:
Since this is a performance optimization, did you also measure how much faster an `iter` in a `BTreeMap` is compared to the `remove`? Both have O(log(n)) worst case, right?

Reply:
Iterating through all the items in the map without doing any re-balancing, and then dropping the whole map at once, should overall be `O(n)`, whereas doing `n` removes amounts to `O(n log(n))`. Also, removing an item means some moving/copying of data around to re-balance the other items, which is avoided here. That's the primary benefit of this optimization.